src/index/base.cpp

```diff
@@ -66,6 +66,28 @@ bool BaseIndex::Init()
         m_best_block_index = FindForkInGlobalIndex(::ChainActive(), locator);
     }
     m_synced = m_best_block_index.load() == ::ChainActive().Tip();
+    if (!m_synced) {
+        bool prune_violation = false;
+        if (!m_best_block_index) {
+            // index is not built yet
+            // make sure we have all block data back to the genesis
+            const CBlockIndex* block = ::ChainActive().Tip();
+            while (block->pprev && (block->pprev->nStatus & BLOCK_HAVE_DATA)) {
+                block = block->pprev;
+            }
+            prune_violation = block != ::ChainActive().Genesis();
+        }
+        // in case the index has a best block set and is not fully synced
+        // check if we have the required blocks to continue building the index
+        else if (!(m_best_block_index.load()->nStatus & BLOCK_HAVE_DATA)) {
```

Comment:
If `m_best_block_index` is not in the main chain, we may need its parent blocks to undo indexing. This should be checked here too.